### PR TITLE
chore: use typescript language service to compile project

### PR DESCRIPTION
**Pull request by DevVersion**
* Uses the TypeScript Language Service to compile the project.
* Allows us to use the `tsconfig.json` with glob patterns.
* Prints diagnostic errors to console within build script.

Recently introduced in Angular Bot as well.

### CODE REVIEW

File: scripts/build.js [hunk 1]
```diff
@@ -26,13 +26,24 @@ let sourceFiles = buildConfig.tsFiles
...
 
-  console.log("Build: Successfully compiled the TypeScript files into ES5.");
-} catch (e) {
-  console.error("Error: An error occurred while compiling the TypeScript files into ES5.");
-  throw e;
-}
+if (!emitResult.emitSkipped) {
+  console.log(`Build: Successfully compiled the project.`);
+} else {
+  console.error('Build: An error occurred while compiling the project.');
```
Comment (crisbeto) [inline scripts/build.js]: It might be useful to print out the destination dir here.

Comment (DevVersion) [inline scripts/build.js]: Which destination when an error occurred? The enhanced ts-compiler doesn't do that as well.

File: scripts/build.js [hunk 2]
```diff
@@ -26,13 +26,24 @@ let sourceFiles = buildConfig.tsFiles
...
+  let errorMessage = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+  console.error(`${diagnostic.file.fileName}@${lineResults.line + 1}: ${errorMessage}`);
+});
 
-  console.log("Build: Successfully compiled the TypeScript files into ES5.");
-} catch (e) {
-  console.error("Error: An error occurred while compiling the TypeScript files into ES5.");
-  throw e;
-}
+if (!emitResult.emitSkipped) {
```
Comment (crisbeto) [inline scripts/build.js]: Very minor nitpick, which we don't really _need_ to address, but IMO it'd be a little easier to understand if we wrote this is as `if (emitResult.emitSkipped) { console.error } else { ... }`.

Comment (DevVersion) [inline scripts/build.js]: A normal JS / TS convention is that the "actual" pass is in the first block. But I'll change it for you ;)

File: scripts/build.js (hunk 1 above)
Comment (crisbeto) [inline scripts/build.js]: Ops, I mean for the success. You can print the `outDir` from the tsconfig.

Comment (DevVersion) [inline scripts/build.js]: That would make more sense. But in general we wan't to keep it pretty consistent with the TS-compiler and it shouldn't be necessary to pint the outDir here.

Comment (crisbeto) [inline scripts/build.js]: This is mostly for our sake. The TS compiler probably doesn't print the location, because you're calling it from the CLI.

Comment (DevVersion) [inline scripts/build.js]: The TS compiler actually checks the AST and runs the language service as same as we do.

Comment (crisbeto) [inline scripts/build.js]: Either way, it's not a big deal. Let's leave it as it is.

**Comment by ThomasBurleson** [comment]
What was wrong with the Node `exec` version?
What is the advantage with using the `tsc` to compile (and dealing with that complexity)?

**Comment by DevVersion** [comment]
> What was wrong with the Node exec version?

Several things weren't good with using `exec`.
- Errors weren't print, because those are repressed by `exec` - No access to `stdout` and `stderr`
- The TSC binary could change (e.g TypeScript 2.0) and it's pretty hard-coded to spawn the TSC binary.
- When using the TSC binary we're not able to use the `tsconfig.json`, because we supplied the files manually (Globs aren't supported in `tsconfig.json`)
- We are able to control the error diagnostic (likely the first point)

> What is the advantage with using the tsc to compile (and dealing with that complexity)?

I think here is some confusion. TSC is meant to be the CLI compiler of typescript (we used before)
- TSC is also using the Language Service as well, but includes special options.
- Every program which transpiles TypeScript files is using the Language Service (and it's not complex)

The advantage here is (as said above), that we could use our `tsconfig.json`, which will be used by `ts-node` as well. This means that we are super consistent (e.g when changing module resolution / target)

**Comment by ThomasBurleson** [comment]
@DevVersion - great explanation. Thx.
> 👍  re: The primary advantage here is that we can use our tsconfig.json, which will be used by ts-node as well.

**Pull request merged**